### PR TITLE
Jason/fix 444

### LINKED ISSUE #444: missing image viewer when reloading a special layout
if there is only one widget docked and the image viewer widget is undocked, and if we save this kind of layout and reload it, the image viewer will be gone and cannot be launched anymore. 
Screen Shot 2019-08-06 at 2 05 38 PM

### PR DESCRIPTION
**Pull request by duidae**
closes #444 & supporting floating widgets

### CODE REVIEW

**Comment by veggiesaurus** [comment]
@duidae I haven't looked at the code yet, but one thing I noticed is that the floating widgets aren't restored to the correct location. This needs to be saved as well.

**Comment by duidae** [comment]
floatingWidgets in widgetsStore only saves min/default width/height, & tried document.getElementById() but get null object, &  checked the FloatingWidgetComponent/FloatingWidgetManagerComponent, there is no such information either, so wondering is there somewhere else we can get those info?

**Comment by kswang1029** [comment]
In my mind, the minimal requirement is having floating widgets to default size and they are not blocking each others.

**Comment by veggiesaurus** [comment]
@duidae, the `Rnd` component has two useful callbacks: `onResizeStop` and `onDragStop`. Use those to keep track of the widget's current position/size via some additional properties (don't need to be `observable` as far as I can see, since they are purely used for keeping track of things) in `WidgetConfig` maybe?

**Comment by duidae** [comment]
@veggiesaurus oh ok Angus thanks for the info! Let me try it

**Comment by duidae** [comment]
@veggiesaurus  @kswang1029  ok, please review, thx

**Comment by veggiesaurus** [comment]
@duidae this looks _almost _ right. Two issues:
- floating widget resizing seems to only work correctly from right and bottom edges now
- Is it possible to retain compatibility with existing layouts that don't have a floating layout? perhaps we need to start versioning our layout config (on a per-layout basis).

File: src/components/FloatingWidget/FloatingWidgetComponent.tsx [hunk 1]
```diff
@@ -83,20 +83,23 @@ export class FloatingWidgetComponent extends React.Component<FloatingWidgetCompo
...
                 className={className}
                 style={{zIndex: this.props.zIndex}}
-                default={{
-                    // Shift by 5 pixels to compensate for 5px CSS margins
-                    x: widgetConfig.defaultX !== undefined ? widgetConfig.defaultX : widgetsStore.defaultFloatingWidgetOffset + 5,
-                    y: widgetConfig.defaultY !== undefined ? widgetConfig.defaultY : widgetsStore.defaultFloatingWidgetOffset,
-                    width: widgetConfig.defaultWidth,
-                    height: widgetConfig.defaultHeight + headerHeight,
-                }}
+                size={{width: widgetConfig.defaultWidth, height: widgetConfig.defaultHeight + headerHeight}}
```
Comment (veggiesaurus) [inline src/components/FloatingWidget/FloatingWidgetComponent.tsx]: are you sure we need to make this a controlled property? Can't we just simply use the updated values from the events when saving and restoring layouts?

Comment (duidae) [reply src/components/FloatingWidget/FloatingWidgetComponent.tsx]: ah sorry use that prop without fully understanding it's meaning...a little bit confusion actually ToT

**Comment by duidae** [comment]
1 fixed;
about 2, I think versioning is needed for backward compatibility, but I dont have much experience about it, my idea is to add a {verions: 1.0} to CARTA_saved_layouts & add corresponding parser function named with version, i.e. layoutConfigParser1.0(), not sure is that enough? @veggiesaurus can u show some tips about it? 
Btw, can we skip supporting existing layouts without floating widget, since this feature is not released yet, so users dont have those layouts & only exists in our testers, can be deleted by typing localStorage.removeItem("CARTA_saved_layouts") in console

File: src/components/FloatingWidget/FloatingWidgetComponent.tsx (hunk 1 above)
Comment (duidae) [reply src/components/FloatingWidget/FloatingWidgetComponent.tsx]: adding update size/position in componentDidUpdate() fixes the problem with default prop

**Pull request merged**